include/pybind11/numpy.h

```diff
@@ -98,31 +121,43 @@ class array : public buffer {
     }
 
     array(const buffer_info &info) {
+        PyObject *arr = nullptr, *descr = nullptr;
+        int ndim = 0;
+        Py_ssize_t dims[32];
         API& api = lookup_api();
-        if ((info.format.size() < 1) || (info.format.size() > 2))
-            pybind11_fail("Unsupported buffer format!");
-        int fmt = (int) info.format[0];
-        if (info.format == "Zd")      fmt = API::NPY_CDOUBLE_;
-        else if (info.format == "Zf") fmt = API::NPY_CFLOAT_;
-
-        PyObject *descr = api.PyArray_DescrFromType_(fmt);
-        if (descr == nullptr)
-            pybind11_fail("NumPy: unsupported buffer format '" + info.format + "'!");
-        object tmp(api.PyArray_NewFromDescr_(
-            api.PyArray_Type_, descr, (int) info.ndim, (Py_intptr_t *) &info.shape[0],
-            (Py_intptr_t *) &info.strides[0], info.ptr, 0, nullptr), false);
-        if (info.ptr && tmp)
-            tmp = object(api.PyArray_NewCopy_(tmp.ptr(), -1 /* any order */), false);
-        if (!tmp)
-            pybind11_fail("NumPy: unable to create array!");
-        m_ptr = tmp.release().ptr();
+
+        // Allocate zeroed memory if it hasn't been provided by the caller.
+        // Normally, we could leave this null for NumPy to allocate memory for us, but
+        // since we need a memoryview, the data pointer has to be non-null. NumPy uses
+        // malloc if NPY_NEEDS_INIT is not set (in which case it uses calloc); however,
+        // we don't have a descriptor yet (only a buffer format string), so we can't
+        // access the flags. The safest thing to do is thus to use calloc.
+        auto buf_info = info;
+        if (!buf_info.ptr)
+            // always allocate at least 1 element, same way as NumPy does it
+            buf_info.ptr = std::calloc(std::max(info.size, 1ul), info.itemsize);
```

Comment:
Re: `1ul`, yea, I guess I was just lazy :)

Re: malloc vs calloc, here's the only piece of code from numpy that refers to `NPY_NEEDS_INIT`:

```cpp
    if (data == NULL) {
        /*
         * Allocate something even for zero-space arrays
         * e.g. shape=(0,) -- otherwise buffer exposure
         * (a.data) doesn't work as it should.
         */
        if (nbytes == 0) {
            nbytes = descr->elsize;
        }
        /*
         * It is bad to have uninitialized OBJECT pointers
         * which could also be sub-fields of a VOID array
         */
        if (zeroed || PyDataType_FLAGCHK(descr, NPY_NEEDS_INIT)) {
            data = npy_alloc_cache_zero(nbytes);
        }
        else {
            data = npy_alloc_cache(nbytes);
        }
        if (data == NULL) {
            PyErr_NoMemory();
            goto fail;
        }
        fa->flags |= NPY_ARRAY_OWNDATA;

    }
```

Realistically, if we never handle object dtypes, we can just use `malloc` instead of `calloc` I guess.

```python
>>> a = np.array(['foo', 'bar'], dtype=object)
>>> memoryview(a).format
'O'
```

Or maybe we can check if there's any `'O'` in the format string, we switch to `calloc`.